staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+// SizeLimitedDiskCache is an implementation of httpcache.Cache that supplements
+// the in-memory map with persistent storage up to a pre-defined maxSize. After
+// the disk cache size breaches the size limit, the oldest files (by file
+// modification times) are evicted until the size limit is satisfied. This cache
+// is meant to be a substitute for "github.com/gregjones/httpcache/diskcache".
+type SizeLimitedDiskCache struct {
+	maxSize     uint64
+	currentSize uint64
+	fileInfos   []fileInfo
+	diskv       *diskv.Diskv
+}
+
+// Get returns the response corresponding to key if present.
+func (c *SizeLimitedDiskCache) Get(key string) (resp []byte, ok bool) {
+	key = keyToFilename(key)
+	resp, err := c.diskv.Read(key)
+	if err != nil {
+		return []byte{}, false
+	}
+	return resp, true
+}
+
+// Set saves a response to the cache as key, and updates the stored
+// fileInfos and currentSize. This function can cause cache evictions.
+func (c *SizeLimitedDiskCache) Set(key string, resp []byte) {
+	key = keyToFilename(key)
+	size := uint64(len(resp))
+	c.trimDiskCache(size)
+	err := c.diskv.WriteStream(key, bytes.NewReader(resp), true)
+	if err == nil {
+		c.appendFileInfo(fileInfo{name: key, size: size, modTime: time.Now()})
```

Comment:
My concern is that this is ignoring the "Transform" things that can happen in Diskv, which we should probably be agnostic to (and should be). So the `key` namespace we have here might not be the same key namespace as `Diskv` uses (if I'm reading Diskv code correctly).

Reply:
The `Transform` (as well as `Compression`) have to be configured when creating `Diskv`. Kubernetes does not configure path tranformations. All cached files appear to be `HTTP_CACHE_DIR/MD5_URL_FILENAME` in a flat hierarchy. There are no subdirectories. The md5 hashed filename that is stored in the cache has nothing to do with `Diskv`; it was created by Greg Jones in his `httpcache/diskcache` above `Diskv`.

Reply:
As discussed offline, we don't have to worry about our abstraction being implicitly tied to the Diskv creation if we create the diskv instance directly here.